BoxOffice [STEP 3] Jason, Lust3r

JasonLee0223 opened the pull request:
# BoxOffice [STEP 3] Jason, Lust3r
안녕하세요! 제이슨(@ehgud0670)! Jason(자손😇), Lust3r입니다!
Step3를 진행하면서 DiffableDataSource와 Snapshot의 관계, async-await의 비동기 처리, UI 업데이트 부분에서 구현 및 
학습 시간이 오래걸리게 되었네요!

API요청 부분까지는 Clean Architecture에 맞춰서 진행하였지만 View단에서는 아직 ViewModel을 적용하는게 익숙치 않아서 
이후 리펙토링을 진행해보려고합니다!

이번 리뷰 잘 부탁드립니다! 감사합니다! 🙇🏻🙇🏻

## 🧑🏼‍💻 작업 목록

### **[ STEP 3 ] 박스오피스 목록 화면 구현**
- [x] Step 2에서 구현한 네트워킹 기능을 통해 실제로 상품목록을 API 서버에 요청하여 불러옵니다.
- [x] 어제의 박스오피스를 볼 수 있는 화면을 구현합니다.
- [x] 리스트를 아래로 잡아끌어서 놓으면 리스트를 새로고침합니다 (당겨서 새로고침)
- [x] 처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.
- [x] 화면 상단에는 날짜를 표기합니다
- [x] 리스트 형태로 박스오피스 정보를 표기합니다
- [x] 박스오피스 정보의 각 열에 표기할 필수정보
    - [x] 맨 왼쪽에는 영화의 현재 등수를 표기합니다
    - [x] 신규 영화면 등수 아래에 신작이라고 표기합니다
    - [x] 기존 영화면 어제와 비교한 등락을 표기합니다
        - [x] 순위 상승 : 빨간 화살표 + 등락 편차
        - [x] 순위 하락 : 파란 화살표 + 등락 편차
        - [x] 변동 없음 : - 표기
    - [x] 해당 일자의 관객수와 누적 관객수를 표기합니다
        - [x] 숫자가 세 자리 이상 넘어가면 ,를 활용하여 읽기 쉽도록 합니다. 예) 10,000


### **Step 3 핵심 경험**
- [x] Safe Area을 고려한 오토 레이아웃 구현
- [x] Collection View의 활용
- [x] Mordern Collection View 활용

## 📚 학습 키워드 및 학습 내용

- [x] DiffableDataSource & Snapshot 활용
- [x] (Test과정) `ListCell`과 `UIView` & `ContentView`의 사용

## 🤔 고민과 해결
### SnapShot의 적용
- `snapshot에` `Section`과 `DailyBoxOffice`를 append했음에도 API 데이터를 불러왔을 때 반영이 되지 않아 UI 구성에 어려움을 겪었습니다. 전반적인 구조를 다시 살펴본 결과 `Task`내에서 API 데이터를 불러온 후, `snapshot` 설정을 하고 `apply()`를 하면 데이터를 출력할 수 있음을 알게 되었습니다.

### ✍🏻 구현 과정
이전 까지는 `UICollectionViewDataSource`를 사용했었는데 이번에는 `DiffableDataSource`를 적용해보려고 노력했습니다.
덧붙여 `ListCell`을 사용해보고싶다는 Lust3r의 의견을 참고해 적용해보려고 하였는데 생각보다 쉽지 않았습니다.
(기존에 `UICollectionViewCell`처럼 사용하면 되는 부분에서 뭔가 새로운 것을 적용한다는 점에서 ContentView를 새로 만들어서 적용해야한다는 부분
또는 Override를 이용하여 재정의를 통해 사용해야할 것 같다는 생각에 갇혀있었던 것 같습니다.)

CollectionView를 구현하는 과정에서는 API를 통해 호출하여 DTO모델과 새로운 Entity를 연결하는 부분에서 networkService를 어디에서 호출하여 사용해야할까?
고민을 많이 하게되었습니다. (viewModel을 사용하게되면 자연스럽게 networkService를 viewModel에서 사용하면 될 것 같다는 점..?)
하여 Task를 호출하는 부분을 메서드로 분리하여 사용하게 되었습니다.


## 💻 실행결과

<img src= "https://user-images.githubusercontent.com/92699723/237051999-486746e6-e3ae-48a3-bf22-1c3f1e2ebdbb.gif" width=30%>

Commit:
[Style] Infrastructure/Network 경로로 Query 그룹 이동

Commit:
[Feat] CollectionView 초기 구현

- collectionView 만들기
- 기존 ViewController -> HomeViewController 네이밍 변경

Commit:
[Feat] navigationBar 구현 및 적용

llimental committed:
[Feat] CollectionViewCell 초기 구현 및 Deployment Target 변경

- BoxOfficeCollectionViewCell.swift
- UICollectionViewListCell을 사용하기 위해 Deployment Target iOS 13.0 -> 14.0 변경

Commit:
[Refactor] 기존 UI 컴포넌트 요소들 클로저 형태로 리펙터링

Commit:
[Style] BoxOfficeListCell로 네이밍 변경

llimental committed:
[Feat] 영화 제목과 관객 수 레이블을 담을 BoxOfficeBrief 클래스 초기 구현

llimental committed:
[Style] 영화 정보 Label 코드를 BoxOfficeBrief 클래스로 이동

Commit:
[Feat] UI 컴포넌트(요소)의 내부 속성 설정 및 메서드 구현

llimental committed:
[Feat] 영화 랭크 정보를 담은 Stack을 구성하는 BoxOfficeRank class 구현

Commit:
[Feat] BoxOfficeListCell의 프로퍼티 구현

llimental committed:
[Feat] datasource에 사용할 section enum타입 구현

Commit:
[Feat]: DailyBoxOffice 엔티티 타입 구현

llimental committed:
[Feat]: CollectionView에 사용할 DiffableDataSource 구현

Commit:
[Feat]: CollectionView Layout 메서드 구현

- ListConfiguration 사용

Commit:
[Style]: 코드 정리

llimental committed:
[Feat]: DiffableDataSource에 사용할 Snapshot 구현

llimental committed:
[Refactor]: BoxOfficeCell 타입을 ListCell에서 Cell로 변경

llimental committed:
[Feat]: BoxOfficeCell Layout 제약 추가

llimental committed:
[Fix]: Cell Layout 이슈 해결

- addArrangedSubView로 인해 두 StackView가 강제적으로 맞춰지기 때문에 다른 제약을 주고자 addSubView로 변경하여 해결

Commit:
[Feat]: networkService를 사용하는 fetchData 메서드 테스트 코드 작성

llimental committed:
[Feat]: API 호출 데이터 UI 출력

llimental committed:
[Refactor]: 사용하지 않는 snapShot 보일러 플레이트 코드 제거

Commit:
[Feat]: 누적 관객 수 프로퍼티 사용을 위해 추가

Commit:
[Feat]: NumberFormatter를 사용하는 메서드 구현

Commit:
[Style]: 필요없는 코드 삭제

Commit:
[Refactor]: UICollectionViewCell -> UICollectionViewListCell로 타입 및 파일명 변경

Commit:
[Feat]: Cell.accessaries 추가

llimental committed:
[Feat]: navigationBar 배경을 맞춰주기 위해 isTranslucent 속성 변경

Commit:
[Refactor]: Layout 수정

llimental committed:
[Fix]: Cell Layout 이슈 해결

Commit:
[Refactor]: movieType -> rankVariation 네이밍 변경

Commit:
[Feat]: rankVariation UI Attribute 변경 메서드 구현

llimental committed:
[Feat]: 신작 혹은 기존 작품 여부를 판별할 rankOldAndNew 프로퍼티 추가

llimental committed:
[Feat]: determineRankVariation 메서드 구현

- rankOldAndNew를 통해 신작 여부 판단
- rankVariation을 통해 0과 그 이외의 값 구분
- 순위 변동 값과 UIColor 값 반환

llimental committed:
[Style]: 텍스트와 이미지 색상을 변경하기 위해 setRankVariation 메서드 수정

llimental committed:
[Feat]: 증감에 따른 UIImage와 색상을 반환하기 위한 determineVariationImage 메서드 구현

Commit:
[Chore]: Optional 타입으로 변경

Commit:
[Chore] '신작'일 때 출력하지 않도록 로직 변경

Commit:
[Feat]: Refresh Control 적용

Commit:
[Feat] Calender를 사용하여 Date날짜 설정하는 메서드 구현

llimental committed:
[Feat]: MagicLiteral enum 구현

llimental committed:
[Style]: MagicLiteral을 사용하여 텍스트 대치

Commit:
[Style] DataMapping -> DataStructure 그룹 네이밍 변경

llimental committed:
[Feat]: formatting 기능을 수행하기 위한 타입 구현

- Interface: Convertible protocol 구현
- Helper: Convertible protocol을 채택한 Formatter 구조체 구현

llimental committed:
[Feat]: 영화 rank 관련 determine 작업을 하기 위한 타입 구현

- Interface: Decidable protocol 구현
- Helper: Decidable protocol을 채택한 DecideHelper 구조체 구현

llimental committed:
[Refactor]: ViewController에 DIP 적용

- ViewController extension을 통해 사용했던 메서드를 분리하여 사용

Commit:
[Chore]: 타입 네이밍 Selector로 변경

Commit:
[Chore]: testEntity -> dailyBoxOfficeStorage로 네이밍 변경

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+        ])
+    }
+
+    private func configureOfCollectionViewLayout() -> UICollectionViewLayout {
+        let configure = UICollectionLayoutListConfiguration(appearance: .plain)
+        let layout = UICollectionViewCompositionalLayout.list(using: configure)
+        return layout
+    }
```
이 메소드는 결국 생성한 Layout 을 반환하는, 즉 리턴값이 있는 메소드이므로 create 나 make 와 같은 메소드가 어울린데 어떠신가요~? 현재 이름인 configure는 메소드 setup 처럼 설정하는 목적의 함수로 반환값이 없는 함수이어서 조금 어색하네요

JasonLee0223 replied:
넵! 네이밍이 추천해주신 단어가 조금 더 직관적인 것 같아서 적용하였습니다! 

해당 커밋: 1700859

ehgud0670 commented on BoxOffice/Presentation/Cells/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,62 @@
+
+class BoxOfficeListCell: UICollectionViewListCell {
+
+    static let identifier = String(describing: BoxOfficeListCell.self)
```
이런건 프로토콜로 빼서 앞으로도 재사용 할 수 있게끔 만드시면 좋을 것 같아요.

JasonLee0223 replied:
잠깐 미팅 때도 언급하긴했지만 이번에 리펙토링하면서 많이 배웠습니다! 감사합니다!!

해당 커밋: d09efa0

ehgud0670 commented on BoxOffice/Presentation/Cells/BoxOfficeBrief.swift:
```diff
@@ -0,0 +1,57 @@
+        return audienceCount
+    }()
+    
+    private func configuration() {
+        axis = .vertical
+        alignment = .leading
+        distribution = .fillEqually
+    }
+    
+    private func configurationOfComponents() {
+        addArrangedSubview(movieName)
+        addArrangedSubview(audienceCount)
+    }
```
음,, 왜 여기만 메소드인데 명사로 하신거죠?

JasonLee0223 replied:
구현 과정에서 급하게 적다보니 놓친 것 같습니다..🥲

해당 커밋: dd37f01

ehgud0670 commented on BoxOffice/Presentation/Cells/BoxOfficeBrief.swift:
```diff
@@ -0,0 +1,57 @@
+import UIKit
+
+final class BoxOfficeBrief: UIStackView {
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        configuration()
+        configurationOfComponents()
+    }
+
+    required init(coder: NSCoder) {
+        super.init(coder: coder)
+        configuration()
+        configurationOfComponents()
+    }
+
+    private let movieName: UILabel = {
+        let movieName = UILabel()
+        movieName.font = .systemFont(ofSize: 20)
+        movieName.textColor = .black
+        return movieName
+    }()
+
+    private let audienceCount: UILabel = {
+        let audienceCount = UILabel()
+        audienceCount.font = .systemFont(ofSize: 15)
+        audienceCount.textColor = .black
+        return audienceCount
+    }()
```
기본적으로 프로퍼티를 클래스 맨 앞에다 두고 그 다음에 init이 와야 돼요. 
프로퍼티를 아래로 뺀 것은 아마 프라이빗한 변수라서 init 다음에 두신걸까요?
그렇다면 프라이빗 메소드 아래에 프라이빗한 프로퍼티들을 모아두는게 보기 더 자연스러울 것 같아요 :)

JasonLee0223 replied:
평소에 위 부분이 정말 많이 헷갈렸는데요! 명확하게 정해주셔서 감사합니다!! 😇
앞으로는 망설이지 않고 잘 작성할 수 있을 것 같아요! ㅎㅎ

해당 커밋: f4e9e44

ehgud0670 replied:
ㅎㅎ 정답은 없는데요, 저도 아래 [아티클](https://soojin.ro/blog/raw-code)을 참고했습니다. 👍👍

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+
+    //MARK: - Private Property
+
+    private lazy var navigationBar : UINavigationBar = {
```
self 변수에 접근하기 위해 lazy 를 사용하셨는데, 채용 과제에서 이 이유 때문에 lazy를 사용하셨으면 반드시 
* lazy 키워드는 무엇인지
* 어느 상황에 사용하는게 적절한지, 어떻게 동작하는지
* 멀티 스레드 환경에 안전한지 

등 여러 질문을 곧이곧대로 물어보는게 아니라 꼬아서 면접관이 여쭤보십니다. 질문 대비가 되어 있으신다면 과제에서 Lazy 키워드를 써도 되고 안되어 있다면 안쓰시는 게 나을 것 같아요

JasonLee0223 replied:
보통 lazy를 사용할 때는 변수가 처음 요청되었을 때만 사용자가 지정한 함수를 사용하여 생성되는 것으로 알아서
메모리 측면에서 좋다고 생각되어 자주 사용되었던 부분이었는데 사용할 때 좀 더 고민을 많이 해보고 사용해야겠습니다!

또한 다중 스레드 환경에서 lazy 프로퍼티에 동시 다발적으로 접근할 때는 한 번만 초기화된다는 보장이 안됨으로
생성되지 않은 lazyh 프로퍼티에 많은 스레드가 비슷한 시점에 접근한다면 여러 번 초기화 될 수 있어서 멀티쓰레드 환경에서의
스케쥴 관리를 생각해야하는데... 아직은 이런 부분을 잘다룰 수 없다고 생각되어 앞으로 사용할 때 조심히 써야겠습니다!

다시 한 번 기본을 되짚어보게 된 것 같습니다! 조언 감사합니다!! 

해당 커밋: 3b8137f 817466a c4da44b

ehgud0670 replied:
음, 기존에 적으신 lazy 객체를 보면 해당 화면에 진입했을 때마다 항상 접근이 되어서 메모리에 항상 할당되는 것 같은데 메모리 측면에서 어느 부분이 무엇이 좋은건가요?

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+        return collectionView
+    }()
+
+    private var dataSource: UICollectionViewDiffableDataSource<Section, DailyBoxOffice>!
```
음 옵셔널 forced unwrapping은 그냥 자제 해주셨으면 합니다. 
정말 나중에 어떤 경우에도 크래시가 안난다고 자신 있을 때 그때 사용하시면 좋을 것 같아요 :)

+ 어떤 경우에는 일부러 `!` 를 사용합니다. 로컬로부터 리소스를 불러올 때 같은 경우 입니다. 예를 들어 Assets으로 부터 이미지 사진을 불러오는 경우처럼요 (ex.`UIImage(named: "Happy")!`) 이유는 배포전에 자신이 로컬에 넣은 리소스가 잘 돌아가는지 테스트하기 위함입니다 :)

ehgud0670 replied:
리소스를 불러올 때 강제 언래핑에 오히려 더 낫다고 하는 글을 찾았습니다. Swift-Book 에서 [Disabling Error Propagation](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/errorhandling/#Disabling-Error-Propagation) 부분입니다. 
다만 전 하나의 의견을 드린거지 정답은 없는 것 같아요.

<img width="1063" alt="스크린샷 2023-05-11 오전 11 29 04" src="https://github.com/tasty-code/ios-box-office/assets/38216027/f7218e3d-1b2f-4e18-b3f7-82ccbbbaf4ff">

JasonLee0223 replied:
위 내용은 처음부터 참 고민이 많았습니다...🥲 
요구사항에서도 명시되어있듯이 강제 언래핑은 사용하지 않는 부분이 있어서 옵셔널로 구현하려고 하였습니다.

그러다 공식문서에 Modern CollectionView Project에서 아래 추가로 코멘트해주신 것처럼 강제 언래핑이 
오히려 더 낫다는 경우를 보여주는 예제 코드를 많이 보게되었습니다. 
해당 내용을 통해 저희 또한 정답은 없을 것 같다고 생각되어 강제 언래핑 되어있는 부분으로 구현하게되었습니다.

항상 코드에 정답은 없지만 제이슨은 어떻게 하는 방법을 좀 더 선호하실까요?? 🤔
(p.s. 이번 프로젝트에서는 기존 코드를 그대로 사용하려고 합니다.)

ehgud0670 replied:
말씀하신대로 정답은 없습니다.. 음.. 저도 그럼 공식문서의 Modern Collectionview Project 를 보고 다시 한번 말씀드릴게요! 👍🙂

ehgud0670 replied:
사실 써놓으신 코드도, 공식문서의 프로젝트 코드도 ViewDidLoad에서 dataSource를 항상 할당하기 때문에 위 코드는 해당 화면에서 제가 위에 드린 코멘트 `정말 나중에 어떤 경우에도 크래시가 안난다고 자신 있을때 그때 사용하시면 될 것 같아요 :)` 의 경우가 맞습니다 :)
그럼에도 `음 옵셔널 forced unwrapping은 그냥 자제 해주셨으면 합니다.` 라고 말씀드렸던 건 
사실 forced unwrapping 을 했을 경우 그리고 nil 인데 접근했을 경우 크래시 발생에 대한 치명성에 대해서 인지하지 못하고 있으셨던 건 아닌지 걱정된 마음에 코멘트 드렸습니다.

하지만 지금 당장은 어떤 경우에도 크래시 안난다고 확신하나 기획이 바뀌어 크래시가 날 수 있는 경우는 위험합니다. 
예를 들어, `!`으로 선언한 객체를 viewWillAppear 에서 할당하는 코드는 위험합니다. 
기획적으로 viewDidLoad 에서 dataSource에 접근하는 경우가 지금 당장은 없더라도 기획이 바뀌는 경우에는 위험할 수 있기 때문입니다. 

* 제가 하는 방법. 
  * 저는 항상 `!` 선언을 지양하는 개발을 해왔습니다. 하지만 무조건적으로 어떤 경우에도 크래시가 나지 않는 객체가 분명히 있습니다.  따라서 그 경우에는 non-optional 로 선언하거나 non-optional 로 선언하는게 번거로울 경우 여러분들처럼 `!` 로 선언할 듯 합니다.
  * 다만 nil error로 인한 크래시 발생은 치명적이니 잘 분간해서 사용하시면 될 것 같습니다 😊👍

JasonLee0223 replied:
자세한 설명 감사합니다! 🙇🏻
지금까지 학습하면서 말씀해주신 것처럼 `정말 어떤 경우에도 크래시가 안난다고 자신 있을 때만 사용하면 된다는 점`이란 부분에 대해
예상치못한 부분의 에러를 확인할 수 없으니 항상 에러 처리에 신경쓰려고 노력했습니다.

다른 리뷰어분들 뿐만 아니라 제이슨이 아래 comment에서도 말씀해주신 것처럼 유지보수 면에서는 공식문서 예제코드보다 
본인 기준으로 고려해서 개발하는 점 이번에 많이 느낀 것 같습니다.

아무래도 학습하는 과정에서는 배우는 입장이다보니 좋은 부분은 모두 흡수하고 싶은 마음에 공식 문서의 중요성을 많이 듣다보니
코드 또한 따라가게되었네요...
이제는 수료까지 얼마 안남았기에 마지막 프로젝트에서만큼은 이제 개발자라는 직업을 가지게 될 사람의 모습에서 
`내가 작성한 코드에 이유가 있고 다른 사람들 또한 이해하고 설득될 수 있게 작성해야겠다`라고 다시 한 번 느끼게 됬습니다.
감사합니다! 👍🏻

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+        let yesterdayDate = formatter.receiveCurrentDate().split(separator: "-").joined()
+        let boxOfficeRequestDTO = BoxOfficeRequestDTO(targetDate: yesterdayDate)
+
+        var networkResult: BoxOfficeResult?
```
이거 필요없는 변수라 빼도 될 것 같아요.

JasonLee0223 replied:
코드 정리 과정에서 잔여 코드가 남았네요...! 삭제하였습니다!

해당 커밋: c5c0c6c

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+        Task {
+            let result = try await networkService.request(with: APIEndPoint.receiveBoxOffice(with: boxOfficeRequestDTO))
+            networkResult = result.boxOfficeResult
+            networkResult?.dailyBoxOfficeList.forEach({ officeList in
+                dailyBoxOfficeStorage.append(DailyBoxOffice(movieBrief: MovieBrief(movieName: officeList.movieName, audienceCount: officeList.audienceCount, audienceAccumulated: officeList.audienceAccumulate), rank: Rank(rank: officeList.rank, rankVariation: officeList.rankVariation, rankOldAndNew: officeList.rankOldAndNew)))
+            })
```
본래 forEach 나 map 등은 순수함수로 써야해서 외부 변수의 값을 변경하면 안됩니다. 따라서 순수 함수가 될 수 있도록 코드 리팩토링 부탁드려용 :)

JasonLee0223 replied:
넵! 조언 감사합니다! 해당 내용을 아래와 같이 수정하였습니다!

```Swift
let temporaryStorage = receivedFromNetworkService.map { dailyBoxOfficeList in
                
    DailyBoxOffice(movieBrief: MovieBrief(movieName: dailyBoxOfficeList.movieName,
                                          audienceCount: dailyBoxOfficeList.audienceCount,
                                          audienceAccumulated: dailyBoxOfficeList.audienceAccumulate),
                   rank: Rank(rank: dailyBoxOfficeList.rank,
                              rankVariation: dailyBoxOfficeList.rankVariation,
                              rankOldAndNew: dailyBoxOfficeList.rankOldAndNew))
}
```

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    @objc func handleRefreshControl() {
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
+            self.refresh.endRefreshing()
+        }
```
음,,, 리프레시를 하면 서버 API를 한번 더 호출해야 할 것 같습니다만,,,

JasonLee0223 replied:
넵 맞습니다..ㅎ 우선적으로 Step3가 너무 질질 끌어지는 느낌이 많이 들어서 피드백을 받고 
해당 API 호출 한 번 더하는 코드를 추가하려고 하였는데 View에 refreshControl이 보여지는 부분만 우선적으로 
구현하게되었습니다...😅 해당 내용은 수정하였습니다!

```Swift
@objc func handleRefreshControl() {
        applySnapshot()
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
            self.refresh.endRefreshing()
        }
    }
```

ehgud0670 commented on BoxOffice/Presentation/Cells/BoxOfficeBrief.swift:
```diff
@@ -0,0 +1,57 @@
+
+import UIKit
+
+final class BoxOfficeBrief: UIStackView {
```
음,,, 변수명 자체가 뷰라는 느낌이 사실 하나도 들지 않습니다. 뷰라는 느낌이 들게 변수명 바꿔주시는게 좋을 것 같아요 ㅎㅎ

JasonLee0223 replied:
최대한 타입명을 사용하지않고 어떻게 표현하면 좋을까? 하는 과정에서 나온 네이밍이였습니다.
Brief 단어가 저희도 애매하였는데 수정하게되어 잘된 것 같습니다!

해당 커밋: a5c9f45

ehgud0670 replied:
아 네이밍을 타입명 없이 할수 있고 저도 그렇게 할 수도 있다고 들었던 것 같아요! 하지만 일단 지금까지 봐왔던 코드도 뷰객체면 타입명(ex. View, Label 등등) 이 있었고,
저의 경우에는 View라는 타입명이 붙어야 직관적으로 구분되어 헷갈리지 않더라고요..
음 계속 시도해보면 좋겠지만 항상 상대방이 모델값이라고 오해는 안할지 생각하면서 객체 네이밍 하시면 될 것 같아요!

ehgud0670 replied:
SummaryInformationView 
이름 매우 좋네요 고생하셨습니다 👍

ehgud0670 commented on BoxOffice/Presentation/Cells/BoxOfficeRank.swift:
```diff
@@ -0,0 +1,84 @@
+
+import UIKit
+
+final class BoxOfficeRank: UIStackView {
```
이것도 뷰가 아니라 모델 값 같아요 RankView 라고 하는게 어떨까요?

JasonLee0223 replied:
넵 위와 동일하게 생각되어 수정하였습니다!

해당 커밋: c8da15e

ehgud0670 commented on BoxOffice/Presentation/Cells/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,62 @@
+
+    override init(frame: CGRect) {
+        super .init(frame: frame)
+        setUpAllUIComponents()
```
왜 딴곳은 configure로 메소드명 쓰셨는데 여기는 왜 setup인가용? 둘 중 하나로 정하는게 통일감 있어보여 좋을 것 같아요.

JasonLee0223 replied:
setUp이라는 단어가 좀 더 직관적이라고 생각되어 사용했었습니다. (동일한 의미이면 역시 통일감이 중요했네요..!! 😅)
해당 내용 수정하였습니다! 

해당 커밋: 81a6ee7

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+
+    // MARK: - Property
+
+    var networkService: NetworkService
+    var dailyBoxOfficeStorage: [DailyBoxOffice]
+    var decideHelper: Decidable
+    var formatter: Convertible
```
여기는 왜 private 으로 두지 않으셨죠? 궁금하네요.

JasonLee0223 replied:
구현 과정에서 놓친 것 같습니다.. 수정하였습니다! 

해당 커밋: 8cb42d8

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+
+import UIKit
+
+class HomeViewController: UIViewController {
```
final을 쓴 곳도 있고 안쓴 곳도 있네요. 쓰든 안쓰든 하나로 정하셔서 통일감있게 작성하셨음 좋겠고, final 을 사용하면 어떤 이점이 있는지도 알고 계셨으면 하네용. 면접 단골 질문입니다.

JasonLee0223 replied:
`final` 키워드를 사용하면 컴파일 과정에서 Override 되지 않는 다는 점을 바로 알아차릴 수 있어서 RC를 생각하는
측면에서 좋다는 것으로 알고있습니다!

파일을 생성하면서 클래스 타입을 사용할 때는 final 키워드 사용이 많이 습관처럼 되었는데 ViewController에 적용하는게
아직 자리가 잘 안잡힌 것 같습니다..ㅠㅠ 
(아마 @llimental (Lust3r)와 merge 과정에서 반영이 안된 부분도 있었을 것 같습니다..ㅠ)

해당 커밋: d89448d

ehgud0670 replied:
오버라이딩도 맞는 말씀입니다! 👏
다만 final 클래스를 사용하는 것과 private function 을 사용하면 성능적으로(Performance) 어떤 이점이 있는지도 알고 계셨으면 좋겠어요!

관련해서 제가 도움이 되었던 링크입니다.
https://www.youtube.com/watch?v=z1Gf6EosaUQ

ehgud0670 replied:
RC를 생각하는 측면에서 무엇이 좋은지는 정확히는 모르겠습니다.. 어떻게 좋은지 말씀 부탁드려용

제가 알고있는건 말씀주신 것처럼 final class 를 사용하면 상속에 닫혀있어 성능적으로 이점이 생긴다는 점입니다.
관련해서 제가 도움이 되었던 링크입니다.
https://www.youtube.com/watch?v=z1Gf6EosaUQ

JasonLee0223 replied:
참고자료 감사합니다! 마침 이전에 학습했던 스위프트 성능 이해하기 내용이였네요! ㅎㅎ
제가 남긴 글에서 `RC를 생각하는 측면`에서 부연설명으로 `컴파일 타임`에 final 키워드가 있음으로서 말씀해주신 것처럼 
상속에 닫혀있어 성능으로 이점이 생긴다는 점이였습니다! 

(정확히 어디서 봤는지 참고 자료를 못찾겠네요...🥲 제 기억으로는 컴파일러가 빌드 시에 final 키워드가 있으면 
컴파일러가 그 class의 내부까지 들어가지 않고 상속에 닫혀있음을 인지하여? 성능에 이점이 있다고 이해했었습니다.)

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+    // MARK: - Property
+
+    var networkService: NetworkService
+    var dailyBoxOfficeStorage: [DailyBoxOffice]
```
이 값은 모델인데 해당 배열 값을 관장하는 상위모델이 있었으면 합니다.
그리고 해당 배열값을 private 하게 감추고 메소드로 동작을 표현하는거죠.

JasonLee0223 replied:
미팅 때 언급해주셨던 것처럼 ViewModel 파일을 만들어서 코드 분리를 많이 하게되었습니다! 감사합니다!

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+
+    //MARK: - Initializer
+
+    init(networkService: NetworkService, testEntity: [DailyBoxOffice], decideHelper: Selector) {
+        self.networkService = NetworkService()
+        self.dailyBoxOfficeStorage = [DailyBoxOffice]()
+        self.decideHelper = Selector()
+        self.formatter = Formatter()
+
+        super.init(nibName: nil, bundle: nil)
+    }
```
음,,, 뷰 컨트롤러가 현재 스토리보드를 쓰고 있는데 이 코드는 왜 구현하신걸까요? 궁금합니다.

JasonLee0223 replied:
미팅 때 잠깐 말씀드렸던 부분입니다!

ViewController에 너무 많은 코드가 생겨서 최대한 분리하기 위해서 위와 같이 초기화를 사용하였습니다.
구현 과정에서 `super.init(nibName: , bundle: )` 메서드가 필요함을 알아서 위와 같이 작성하게되었습니다.

책임 분리과정에서 해당 코드는 삭제되었습니다!

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+
+        var networkResult: BoxOfficeResult?
+
+        Task {
+            let result = try await networkService.request(with: APIEndPoint.receiveBoxOffice(with: boxOfficeRequestDTO))
+            networkResult = result.boxOfficeResult
+            networkResult?.dailyBoxOfficeList.forEach({ officeList in
+                dailyBoxOfficeStorage.append(DailyBoxOffice(movieBrief: MovieBrief(movieName: officeList.movieName, audienceCount: officeList.audienceCount, audienceAccumulated: officeList.audienceAccumulate), rank: Rank(rank: officeList.rank, rankVariation: officeList.rankVariation, rankOldAndNew: officeList.rankOldAndNew)))
+            })
+            applySnapshot()
+        }
```
현재
* NetworkService 를 이용해 API 호출해 모델값 가져오기 -> 해당 모델 값을 UI에 적용하기
코드가 함께 있는데 분리해야 합니다.

> to-be
(1) 옵저버 패턴을 이용해 뷰나 뷰컨트롤러는 모델의 변화를 Observing 하고 
(2) 모델은 API 통신처럼 모델의 변화가 있을때 Notification을 보내 뷰에게 알려야 합니다. 
(3) 그럼 뷰는 해당 노티를 받고(옵저빙을 했으니 노티를 받을 수 있습니다) 알아서 뷰를 업데이팅해야 합니다.

JasonLee0223 replied:
조언해주신 방법으로 NotificationCenter를 사용하여 데이터를 받아오도록 리펙토링하였습니다!!
감사힙니다!!

llimental replied:
+ NotificationCenter를 활용해보았는데, 저희 코드와 같이 데이터를 업데이트 할 때 그 과정이 끝나면 post를 하고 데이터를 적용하는 쪽에서 Observer로 캐치해서 적용한 것이 옳게 구현된 방식일까요???

ehgud0670 replied:
네, 근데 보통 뷰컨트롤러에서 addObserver 하고 모델데이터를 가지고 있는 뷰모델에서 post 를 보냅니다 :)

llimental replied:
감사합니다!
NetworkService(post) -> ViewModel(observer) -> ViewController(closure) 의 방식에서
NetworkService(closure) -> ViewModel(post) -> ViewController(observer)의 방식으로 변경하였고,
기존에는 observer에서 바로 클로저로 받은 데이터를 사용하는 방식을 사용했었는데 옵저버 등록 따로 하고, selector를 통해 notification 받은 것을 타입캐스팅 하여 스냅샷에 적용하는 것으로 로직을 변경했습니다!

해당 커밋: e5f836b, b735426, 204bdf0

ehgud0670 replied:
확인했습니다 👍
하나 더 첨언하자면 NetworkService의 loadData 메소드도 async 메소드로 만들어도 될 것 같아요 ㅎㅎ
그럼 고생하셨어요!

ehgud0670 commented on BoxOffice/Presentation/HomeViewController.swift:
```diff
@@ -0,0 +1,166 @@
+    func configureDataSource() {
+
+        let cellRegistration = UICollectionView.CellRegistration<BoxOfficeListCell, DailyBoxOffice> { [self] (cell, indexPath, dailyBoxOffice) in
+            let rankVariation = decideHelper.determineRankVariation(with: dailyBoxOffice.rank.rankVariation, and: dailyBoxOffice.rank.rankOldAndNew)
+            let rankImage = decideHelper.determineVariationImage(with: dailyBoxOffice.rank.rankVariation)
+            
+            cell.boxOfficeBrief.setMovieName(by: dailyBoxOffice.movieBrief.movieName)
+            cell.boxOfficeBrief.setAudienceCount(by: formatter.convertToNumberFormatter(dailyBoxOffice.movieBrief.audienceCount,
+                                                                                   accumulated: dailyBoxOffice.movieBrief.audienceAccumulated))
+            cell.boxOfficeRank.setRankVariation(by: rankVariation.0, with: rankVariation.1)
+            
+            if dailyBoxOffice.rank.rankOldAndNew == RankOldAndNew.new || dailyBoxOffice.rank.rankVariation == MagicLiteral.zero {
+                cell.boxOfficeRank.setRankVariation(by: nil, with: nil)
+            } else {
+                cell.boxOfficeRank.setRankVariation(by: rankImage.0, with: rankImage.1)
+            }
+            
+            cell.boxOfficeRank.setRank(by: dailyBoxOffice.rank.rank)
+            
+            cell.accessories = [.disclosureIndicator()]
```
이 코드를 셀에 옮겨 한줄로 표현되겠끔 리팩토링 해보세요.

JasonLee0223 replied:
DiffableDataSource를 구현하는 과정에서 공식문서 예제 코드가 위와 같이 되어있어 위와 같이 사용하였는데
분리하여 훨씬 코드가 클린해졌습니다! 조언 감사합니다! 

```Swift
let cellRegistration = UICollectionView.CellRegistration<BoxOfficeListCell, DailyBoxOffice> { (cell, _ , dailyBoxOffice) in
    cell.configureOfCellRegistration(with: dailyBoxOffice)
}
```

해당 커밋: 2d49a39

ehgud0670 replied:
고생하셨습니다 :) 
유지보수 관련해서는 공식문서 말고 본인 기준 유연성 확장성 가독성 등등 고려해서 개발하시면 좀 더 큰 성장이 있으실겁니다!

JasonLee0223 replied:
넵! 조언 감사합니다! 👍🏻

ehgud0670 commented on BoxOffice/Presentation/Interface/Decidable.swift:
```diff
@@ -0,0 +1,13 @@
+import UIKit
+
+protocol Decidable {
+    func determineRankVariation(with rankVariation: String, and rankOldAndNew: RankOldAndNew) -> (String, UIColor)
+    func determineVariationImage(with rankVariation: String) -> (UIImage, UIColor)
```
* 리턴값이 튜플이라는 건 결국 두가지 역할을 하고 있다는 뜻 입니다. 각각 2개의 메소드가 되도록 분리해보세요(총 4개의 메소드)
* 리턴값이 튜플이면 사용하는 쪽에서 볼때 굉장히 가독성이 좋지 않습니다. 0과 1이 무엇인지 직접 확인해봐야 하기 때문입니다. 🥲

JasonLee0223 replied:
튜플을 사용해보자고 제가 추천을 했네요..ㅎㅎ 

클린 코드 책을 읽으면서 메서드(함수)는 3개 미만의 파라미터(인자)를 가지며 하나의 일만 수행하도록 구현해야한다는
내용을 인지하고 있었지만 함수의 응집도? 측면에서 고민해보았을 때 튜플을 사용하여 하나의 UI 컴포넌트의 속성을 정의하는게
낫지 않을까? 하는 고민에서 시작되었습니다.

말씀해주신 부분에서처럼 0과 1을 사용하면 가독성이 매우 떨어져서 잘 사용하지 않게 되었는데 집어주셔서 감사합니다!
앞으로는 좀 더 명확하게 메서드를 구현할 수 있을 것 같습니다! 감사합니다!!

해당 커밋: c9f0de6

ehgud0670 reviewed (changes requested):
고생하셨습니다! 확인하시고 코멘트 반영 부탁드릴게요
이해가 안된다거나 의문스러운 점은 언제나 말씀부탁드려요~!

ehgud0670 commented:
아, 추가로 구현하셨다고 한 항목들 중에서 
`처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.`
부분은 아직 미구현 상태인것 같은데 확인해주세요~

llimental committed:
[Refactor]: RequestDTO 객체명 QueryParameters로 변경

llimental committed:
[Refactor]: 줄바꿈을 통한 EndPoint의 파라미터 가독성 증진

llimental committed:
[Feat]: APIKEY 코드에서 숨김 처리 및 .gitignore 설정

llimental committed:
[Refactor]: APIMagicLiteral에서 APIKEY 값 제거

llimental committed:
[Refactor]: 은닉한 APIKEY를 불러와 사용할 수 있도록 코드 리팩터링

Commit:
[Feat]: APIKEY 파일 생성 및 키 설정

Commit:
Merge branch 'Refactoring-Step2' of https://github.com/llimental/ios-box-office into Refactoring-Step2

Commit:
[Chore]: 앞에 바꾸지 못한 네이밍 적용

Commit:
[Refactor]: getAPIKEY 메서드 구현 및 적용

Commit:
[Feat]: 프로젝트 타겟에 XCconfig 설정

llimental committed:
[Chore]: configureOfCollectionViewLayout 메서드명 변경

- configureOfCollectionViewLayout() -> createCollectionViewLayout()

llimental committed:
[Feat]: BoxOfficeListCell의 identifier를 protocol로 분리하여 재사용성 증진

Commit:
[Refactor]: 명사로 사용된 메서드명 동사로 변경

Commit:
[Style]: private property를 private method 아래로 코드 위치 이동

Commit:
[Style]: 필요없는 변수 삭제

Commit:
[Chore]: BoxOfficeBrief -> SummaryInforamtionView로 네이밍 변경

Commit:
[Chore]: BoxOfficeRank -> RankView로 네이밍 변경

Commit:
[Chore]: 이전 커밋으로 네이밍 변경된 내용 적용

llimental committed:
[Chore]: 바뀐 view 네이밍에 맞게 프로퍼티명 변경

llimental committed:
[Chore]: setUpAllUIComponents 메서드 네이밍 변경

- setUpAllUIComponents -> configureOfAllUIComponents

Commit:
[Chore]: 기존 public 변수에 private 접근제어 추가 및 private method 분리

Commit:
[Chore]: final 키워드 적용 안되어있던 부분 추가

llimental committed:
[Feat]: SceneDelegate에 window rootViewController 설정

llimental committed:
[Refactor]: HomeViewController의 navigationBar 구현방법 변경

llimental committed:
[Feat]: NavigationBar 위 화면 표현을 위한 코드 수정

llimental committed:
[Feat]: collectionView 구현 방식 변경

Commit:
[Style]: 메서드 코드 위치 이동

Commit:
[Refactor]: receiveCurrentDate 메서드를 stored property로 변경

llimental committed:
[Style]: 필요 없는 코드 제거

llimental committed:
[Refactor]: Selector 메서드 리팩터링을 통한 역할 분리 및 Decidable 프로토콜 반영

llimental committed:
[Refactor]: Selector 메서드 분리에 따른 RankView 메서드 수정

llimental committed:
[Refactor]: 분리한 Selector 메서드를 cell에 사용하도록 리팩터링

Commit:
[Chore]: 인스턴스 생성 코드 위치 변경

Commit:
[Chore]: Presentation 내부 그룹 변경

- UserInterface
   - Cells
   - View
- ViewModel
   - Helper
   - Interface

Commit:
[Refactor]: 기존의 var -> static var로 변경

Commit:
[Refactor]: receiveCurrentDate 타입 변경 내용 적용

Commit:
[Refactor]: postToViewModel 메서드명 loadData로 변경

Commit:
[Style]: 책임 분리로 인해 사용 되지않는 코드 삭제

Commit:
[Feat]: Notification.Name extension으로 정적 상수 추가

Commit:
[Feat]: ViewModel 구현

- initializer
- transformIntoDailyBoxOffice 메서드

Commit:
[Refactor]: networkResult를 받을 수 있도록 메서드 구현 및 적용

Commit:
[Refactor]: BoxOfiiceViewModel 비동기 코드로 수정

Commit:
[Refactor]: dataSource apply 코드 수정

Commit:
[Feat]: 내부 호출만 사용되어 private 접근 권한 추가

Commit:
[Style]: 코드 위치 이동 및 #pragma 추가

Commit:
[Feat]: DailyBoxOffice 타입에 uuid 타입 추가

Commit:
[Refactor]: 기존에 Convertible에 있던 메서드를 Gettable로 옮긴 후 적용

Commit:
[Refactor]: uuid -> identifier 네이밍 변경 및 필요없는 코드 삭제

Commit:
[Refactor]: convertible의 receviceCurrentDate 프로퍼티 코드 이동

Commit:
[Refactor]: Decidable 메서드 모두 extension을 사용하여 프로토콜 기본구현으로 변경

Commit:
[Refactor]: Cell을 통해 UI Component Attribute를 정의하는 부분 코드 이동

Commit:
[Feat]: Converter 파일 생성 및 타입 구현

Commit:
[Style]: 코드 정리

Commit:
[Feat]: activityIndicator 기능 구현 추가

Commit:
[Refactor] ConfigureUIComponents 메서드 구현 및 적용

Commit:
[Style]: 코드 정리

Commit:
[Style]: 개행 삭제

Commit:
[Chore]: viewModel -> boxOfficeViewModel 네이밍 변경

JasonLee0223 commented:
저희가 구현 과정에서 해당 내용을 빼먹었네요...ㅠㅠ
리펙토링 과정에서 해당 내용을 추가하였습니다! 
(activityIndicator를 사용하여 구현하였습니다!)

![Simulator Screen Recording - iPhone 14 Pro - 2023-05-13 at 17 44 15](https://github.com/tasty-code/ios-box-office/assets/92699723/febc0e88-0164-4188-b215-8ea5195b2368)

P.S.
Comment를 제가 혼자 다 남겨서 오해하실 수 있을 것 같아서 추신으로 남깁니다!
해당 구현 과정은 @llimental (Lust3r)와 같이 고민하여 리펙토링 진행하며 미리 적어놓고 조금 남은 구현내용을 주말에 구현하고
올리다보니 제가 답글을 모두 남기게 되었습니다..😅 (러스터는 주말에 일이 좀 있었습니다.)

주말에 알림와서 신경쓰이실 것 같아서 조심스럽네요...! 편하신 시간에 확인 부탁드립니다!! 감사합니다!!!
(refreshControl도 lazy를 제거하려고 하는데 시간이 없어서.. 추가적으로 리펙토링하도록 해보겠습니다! )

llimental committed:
[Feat]: 순위 변동이 없을 때 분기처리하여 출력하도록 구현

ehgud0670 reviewed (changes requested):
<img width="272" alt="스크린샷 2023-05-15 오후 3 40 50" src="https://github.com/tasty-code/ios-box-office/assets/38216027/ca1b5799-d6ec-44b1-b6c2-14bfba1af34f">

현재 API Key Configuration 관련하여 문제가 있나봅니다. 컴파일 에러가 나고 있어요.
Fix 해주시고 리뷰 재요청해주세요. 다시 확인하도록 하겠습니다.

ehgud0670 commented:
혹시 몰라 프로젝트 삭제 후 다시 git clone 했는데도 컴파일 에러 발생하네요~ 한번 확인 부탁드려요~

Commit:
[Refactor]: config 파일 생성

Commit:
[Refactor]: BoxOffice+APIKEY xcconfig파일 추가

Commit:
[Refactor]: XCconfig 파일 추적 금지

Commit:
[Refactor]: 이전 커밋 reset

Commit:
[Refactor]: xcconfig 파일 설정

Commit:
[Refactor]: 프로젝트 xcconfig reset

Commit:
[Chore]: reset을 위해 xcconfig 파일 삭제

Commit:
[Refactor]: APIKEY xcconfig 파일 추가

llimental committed:
[Docs]: Update README.md

llimental committed:
[Refactor]: NetworkService의 loadData 메서드 로직 수정

- loadData의 networkResult를 escaping closure를 통해 전달하도록 변경

llimental committed:
[Refactor]: BoxOfficeViewModel 로직 변경

- NetworkService로부터 받은 데이터를 가지고 매핑 후 NotificationCenter를 통해 ViewController에 post

llimental committed:
[Refactor]: HomeViewController에 옵저버 추가 및 스냅샷 로직 변경

- 옵저버를 추가하여 ViewModel에서 보내는 Notification 수신
- 수신한 데이터를 가지고 applySnapshot() 메서드 수행
